### ISSUE TITLE
Peculiar pathfinding with self-propelled sprayers (CP won't stay on the field)

### ISSUE THREAD

**Issue by SteelRodent** [Courseplay/courseplay#5069]
Having issues with CP failing to keep the self-propelled sprays on the field, so the booms hit trees and stuff on the sides. Tried testing with the new reverse pathfinding in 397 turned on, but that results in the sprayer coming to a full stop and the game freezing (completely unresponsive) for several seconds every time CP has to calculate a turn - which is so annoying that it's pretty much unplayable. With only forward pathfinding on I get really odd turn maneuvers like this:
![20200317030009_1](https://user-images.githubusercontent.com/20990941/76848804-98de6d00-6844-11ea-8063-c032cbd155c2.jpg)
I know the pathfinding code is really complex, but this is just a bit more creative than necessary.

With either pathfinding on it's a bit random whether CP does a nice U-turn or goes into some serious swerving to find the next lane. But either way it won't persistently stay on the field, so it keeps hitting trees.
Was testing on the Pineapple Bay 4x map because I figured the bigger fields would be better with the big sprayers, but it apparently makes no difference. It still eventually turns off the field. 

Was thinking maybe it'd help if it calculates the turns as if the sprayers are much bigger than they are, since the tip of booms are so far from the vehicle, but I understand that would make the code even more complex and possibly not solve the problem anyway.

Towed sprayers still work mostly troublefree, so it's not a big deal that the self-proppelled are misbehaving, but it would be nice to be able to use them. But might want to look into this in the future if you ever have time.

**Comment by Tensuko** [comment]
The long calculation is due to the large hitbox that it got.
Driving outside of the field (at least hit objects with the sprayer arms) is a problem with is not fully integrated yet, specialy at not turns with on offset.
However, with one preturn it should be able to stay on field, normaly.

For now I suggest you to make at least one pre turn and activate only Use AI turn manuvers and deactivate the 2 pathfind options. As this is per vehicle you can still use the other options on other Vehicles, but for the spray I would go for AI Turn for now.

But we will keep an eye on this one.

**Comment by pvaiko** [comment]
@SteelRodent first of all, have you tried turning the 'Use pathfinding in turns' option off? With the course and field layout on your screenshot, there seems to be no need to use pathfinding for the turns at all.

Anyway, that loop does not seem to be right, if you have a savegame, I'll look into that,

**Comment by Crazysonoran** [comment]
I have similar issues with the sprayers mounted on the back of my tractor, but in this case what it does (not every time though) is it turns the wheels one direction, drives back and forth once, turns another direction then drives back and forth again then it attempts to line up for the next leg of it's path and continue on. Sometimes this gets the arm of the sprayer into trees or fences. A lot of times it also takes really wide (off field) paths to get to the next section of the course it needs to follow, cant do 2 headlands sometimes due to smaller fields. I've turned AI and pathfinding on and off and it still does it no matter what combination I try.

**Comment by SteelRodent** [comment]
Sorry I completely forgot I posted this. No longer have the save and it's been so long I can't even remember why I was using the pathfinding at that time. No longer have the save, or even the map, so can't really redo the experiment. 

On Lapacho, however, I find that with the latest versions of CP, generating the course for the Hardi Rubicon with corners set to "turn" (on almost perfectly rectangular or L-shaped fields) makes it do a reverse 270 degree turn in the 90 degree headland corners. which results in a huge square in each corner that isn't worked (since the Rubicon is 48.5 meter it adds up to a pretty big missed area). Generating the course with "round" corners works much better. But I did this with the pathfinding and AI turned off.
